graspy/match/faq.py

```diff
@@ -0,0 +1,250 @@
+        stochastic matrix
+
+    max_iter : int, positive (default = 30)
+        Integer specifying the max number of FW iterations.
+        FAQ typically converges with modest number of iterations
+
+    shuffle_input : bool (default = True)
+        Gives users the option to shuffle the nodes of A matrix to avoid results
+        from inputs that were already matched
+
+    eps : float (default = 0.1)
+        A positive, threshold stopping criteria such that FW continues to iterate
+        while Frobenius norm of (P[i]-P[i+1]) > eps
+
+    Attributes
+    ----------
+    
+    perm_inds_ : array, size (n,) where n is the number of vertices in the graphs
+        fitted.
+        The indices of the optimal permutation on the nodes of B, found via
+        FAQ, to best minimize the objective function f(P) = trace((A^T)PB(P^T)).
+
+        
+    score_ : float
+        The objective function value of for the optimal permutation found.
+        
+
+    References
+    ----------
+    .. [1] J. T. Vogelstein, J. M. Conroy, V. Lyzinski, L. J. Podrazik, S. G. Kratzer,
+    E. T. Harley, D. E. Fishkind, R. J. Vogelstein, and C. E. Priebe, “Fast
+    approximate quadratic programming for graph matching,” PLOS one, vol. 10, no. 4,
+    p. e0121002, 2015.
+
+    """
+
+    def __init__(
+        self,
+        n_init=1,
+        init_method="barycenter",
+        max_iter=30,
+        shuffle_input=True,
+        eps=0.1,
+    ):
+
+        if n_init > 0 and type(n_init) is int:
+            self.n_init = n_init
+        else:
+            msg = '"n_init" must be a positive integer'
+            raise TypeError(msg)
+        if init_method == "rand":
+            self.init_method = "rand"
+        elif init_method == "barycenter":
+            self.init_method = "barycenter"
+            self.n_init = 1
+        else:
+            msg = 'Invalid "init_method" parameter string'
+            raise ValueError(msg)
+        if max_iter > 0 and type(max_iter) is int:
+            self.max_iter = max_iter
+        else:
+            msg = '"max_iter" must be a positive integer'
+            raise TypeError(msg)
+        if type(shuffle_input) is bool:
+            self.shuffle_input = shuffle_input
+        else:
+            msg = '"shuffle_input" must be a boolean'
+            raise TypeError(msg)
+        if eps > 0 and type(eps) is float:
+            self.eps = eps
+        else:
+            msg = '"eps" must be a positive float'
+            raise TypeError(msg)
+
+    def fit(self, A, B):
+        """
+        Fits the model with two assigned adjacency matrices
+        
+        Parameters
+        ---------
+        A : 2d-array, square, positive
+            A square, positive adjacency matrix
+            
+        B : 2d-array, square, positive
+            A square, positive adjacency matrix
+
+        Returns
+        -------
+        
+        self : returns an instance of self
+        """
```

Comment:
https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_array.html

this is probably a better way of checking input (and do `copy=True` `ensure_2d=True`) because it handles a lot more edge cases than we want to think about and make a copy of the inputs so you don't actually modify the input objects accidentally

Reply:
for example, I forgot about checking whether these are 2d arrays

Reply:
@asaadeldin11 did you see this

Reply:
I still do not see this change @asaadeldin11 am i missing it?

Reply:
so should i delete lines 148-156 and instead add A = check_array(A, copy=True, ensure_2d = True)?

Reply:
not quite, still just need to make sure that they are the same shape and square, right? I meant just to make sure they are 2d, make a copy, and are nd.arrays are all handled by check_array. Do you have tests for all of the above?

Reply:
@asaadeldin11

Reply:
@asaadeldin11  not addressed?

Reply:
@bdpedigo  ok. so i'll make the first lines of fit():
A = check_array(A, copy=True, ensure_2d=True)
B = check_array(B, copy=True, ensure_2d=True)
Followed by the shape and square checks